File: static/js/reports.js
/**
 * Reports JavaScript
 * Handles the reporting functionality and data visualization
 */

// Reports state
const ReportsState = {
    salesSummary: null,
    categoryData: null,
    topProducts: null,
    inventoryStatus: null,
    currentPeriod: 'today',
    charts: {}
};

// DOM Elements
document.addEventListener('DOMContentLoaded', function() {
    if (document.getElementById('reportsPage')) {
        initializeReports();
    }
});

/**
 * Initialize the reports page
 */
async function initializeReports() {
    try {
        // Set up event listeners
        setupReportsEventListeners();
        
        // Load initial reports data
        await loadReportsData('today');
        
        // Render reports
        renderReports();
    } catch (error) {
        console.error('Error initializing reports:', error);
        showToast('Error', 'Failed to initialize reports', 'error');
    }
}

/**
 * Set up event listeners for reports page
 */
function setupReportsEventListeners() {
    // Period selector
    const periodButtons = document.querySelectorAll('.period-btn');
    periodButtons.forEach(button => {
        button.addEventListener('click', function() {
            const period = this.getAttribute('data-period');
            loadReportsData(period);
            
            // Update active state
            periodButtons.forEach(btn => btn.classList.remove('active'));
            this.classList.add('active');
        });
    });
    
    // Custom date range form
    const dateRangeForm = document.getElementById('dateRangeForm');
    if (dateRangeForm) {
        dateRangeForm.addEventListener('submit', function(e) {
            e.preventDefault();
            const startDate = document.getElementById('startDate').value;
            const endDate = document.getElementById('endDate').value;
            
            if (startDate && endDate) {
                loadReportsData('custom', startDate, endDate);
            } else {
                showToast('Error', 'Please select both start and end dates', 'error');
            }
        });
    }
    
    // Inventory threshold form
    const thresholdForm = document.getElementById('thresholdForm');
    if (thresholdForm) {
        thresholdForm.addEventListener('submit', function(e) {
            e.preventDefault();
            const threshold = document.getElementById('lowStockThreshold').value;
            loadInventoryStatus(threshold);
        });
    }
    
    // Tab navigation
    const tabButtons = document.querySelectorAll('.nav-link[data-bs-toggle="tab"]');
    tabButtons.forEach(button => {
        button.addEventListener('shown.bs.tab', function(e) {
            // Resize charts when tab is shown
            Object.values(ReportsState.charts).forEach(chart => {
                if (chart) {
                    chart.resize();
                }
            });
        });
    });
}

/**
 * Load reports data for a specific period
 * @param {string} period - Period to load (today, week, month, custom)
 * @param {string} startDate - Start date for custom period (YYYY-MM-DD)
 * @param {string} endDate - End date for custom period (YYYY-MM-DD)
 */
async function loadReportsData(period, startDate = null, endDate = null) {
    try {
        // Show loading state
        AppState.isLoading = true;
        updateLoadingState();
        
        // Update current period
        ReportsState.currentPeriod = period;
        
        // Build URL parameters
        let params = `period=${period}`;
        if (period === 'custom' && startDate && endDate) {
            params += `&start_date=${startDate}&end_date=${endDate}`;
        }
        
        // Fetch data in parallel
        const [summaryRes, categoryRes, productsRes] = await Promise.all([
            apiCall(`/reports/sales/summary?${params}`),
            apiCall(`/reports/sales/by-category?${params}`),
            apiCall(`/reports/sales/top-products?${params}`),
        ]);
        
        // Update state
        ReportsState.salesSummary = summaryRes.data;
        ReportsState.categoryData = categoryRes.data;
        ReportsState.topProducts = productsRes.data;
        
        // Also load inventory status
        await loadInventoryStatus();
        
        // Render reports
        renderReports();
        
        // Update period display
        updatePeriodDisplay();
        
        // Show success message
        showToast('Success', 'Reports data loaded successfully', 'success');
    } catch (error) {
        console.error('Error loading reports data:', error);
        showToast('Error', 'Failed to load reports data', 'error');
    } finally {
        // Hide loading state
        AppState.isLoading = false;
        updateLoadingState();
    }
}

/**
 * Load inventory status
 * @param {number} threshold - Low stock threshold
 */
async function loadInventoryStatus(threshold = 10) {
    try {
        const response = await apiCall(`/reports/inventory/status?low_stock=${threshold}`);
        ReportsState.inventoryStatus = response.data;
        renderInventoryStatus();
    } catch (error) {
        console.error('Error loading inventory status:', error);
    }
}

/**
 * Update the period display
 */
function updatePeriodDisplay() {
    const periodDisplay = document.getElementById('currentPeriod');
    if (!periodDisplay || !ReportsState.salesSummary) return;
    
    const period = ReportsState.salesSummary.period;
    
    let periodText = '';
    switch (period.name) {
        case 'today':
            periodText = 'Today';
            break;
        case 'week':
            periodText = 'Last 7 Days';
            break;
        case 'month':
            periodText = 'This Month';
            break;
        case 'custom':
            const start = new Date(period.start);
            const end = new Date(period.end);
            periodText = `${start.toLocaleDateString()} to ${end.toLocaleDateString()}`;
            break;
        default:
            periodText = 'Custom Period';
    }
    
    periodDisplay.textContent = periodText;
}

/**
 * Render all reports
 */
function renderReports() {
    renderSalesSummary();
    renderSalesByCategory();
    renderTopProducts();
    renderInventoryStatus();
}

/**
 * Render sales summary
 */
function renderSalesSummary() {
    if (!ReportsState.salesSummary) return;
    
    const data = ReportsState.salesSummary;
    
    // Update summary cards
    document.getElementById('totalTransactions').textContent = data.sales.total_transactions;
    document.getElementById('totalSales').textContent = formatCurrency(data.sales.total_sales);
    document.getElementById('averageSale').textContent = formatCurrency(data.sales.average_sale);
    
    // Render payment methods chart
    renderPaymentMethodsChart(data.payment_methods);
}

/**
 * Render payment methods chart
 * @param {Array} paymentData - Payment methods data
 */
function renderPaymentMethodsChart(paymentData) {
    const chartCanvas = document.getElementById('paymentMethodsChart');
    if (!chartCanvas || !paymentData || paymentData.length === 0) return;
    
    // Destroy existing chart
    if (ReportsState.charts.paymentMethods) {
        ReportsState.charts.paymentMethods.destroy();
    }
    
    // Prepare data
    const labels = paymentData.map(item => item.method.toUpperCase());
    const counts = paymentData.map(item => item.count);
    const totals = paymentData.map(item => item.total);
    
    // Create chart
    ReportsState.charts.paymentMethods = new Chart(chartCanvas, {
        type: 'bar',
        data: {
            labels: labels,
            datasets: [
                {
                    label: 'Transaction Count',
                    data: counts,
                    backgroundColor: 'rgba(75, 192, 192, 0.5)',
                    borderColor: 'rgba(75, 192, 192, 1)',
                    borderWidth: 1
                },
                {
                    label: 'Total Amount',
                    data: totals,
                    backgroundColor: 'rgba(54, 162, 235, 0.5)',
                    borderColor: 'rgba(54, 162, 235, 1)',
                    borderWidth: 1,
                    yAxisID: 'y1'
                }
            ]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            scales: {
                y: {
                    beginAtZero: true,
                    title: {
                        display: true,
                        text: 'Transaction Count'
                    }
                },
                y1: {
                    beginAtZero: true,
                    position: 'right',
                    title: {
                        display: true,
                        text: 'Amount'
                    },
                    grid: {
                        drawOnChartArea: false
                    },
                    ticks: {
                        callback: function(value) {
                            return formatCurrency(value);
                        }
                    }
                }
            },
            plugins: {
                tooltip: {
                    callbacks: {
                        label: function(context) {
                            let label = context.dataset.label || '';
                            if (label) {
                                label += ': ';
                            }
                            if (context.datasetIndex === 1) {
                                label += formatCurrency(context.raw);
                            } else {
                                label += context.raw;
                            }
                            return label;
                        }
                    }
                }
            }
        }
    });
}

/**
 * Render sales by category
 */
function renderSalesByCategory() {
    if (!ReportsState.categoryData) return;
    
    const data = ReportsState.categoryData;
    
    // Render categories chart
    renderCategoriesChart(data.categories);
}

/**
 * Render categories chart
 * @param {Array} categoriesData - Categories data
 */
function renderCategoriesChart(categoriesData) {
    const chartCanvas = document.getElementById('categoriesChart');
    if (!chartCanvas || !categoriesData || categoriesData.length === 0) return;
    
    // Destroy existing chart
    if (ReportsState.charts.categories) {
        ReportsState.charts.categories.destroy();
    }
    
    // Prepare data
    const labels = categoriesData.map(item => item.category);
    const sales = categoriesData.map(item => item.total_sales);
    const quantities = categoriesData.map(item => item.quantity_sold);
    
    // Create chart
    ReportsState.charts.categories = new Chart(chartCanvas, {
        type: 'bar',
        data: {
            labels: labels,
            datasets: [
                {
                    label: 'Sales Amount',
                    data: sales,
                    backgroundColor: 'rgba(255, 99, 132, 0.5)',
                    borderColor: 'rgba(255, 99, 132, 1)',
                    borderWidth: 1
                },
                {
                    label: 'Quantity Sold',
                    data: quantities,
                    backgroundColor: 'rgba(153, 102, 255, 0.5)',
                    borderColor: 'rgba(153, 102, 255, 1)',
                    borderWidth: 1,
                    yAxisID: 'y1'
                }
            ]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            scales: {
                y: {
                    beginAtZero: true,
                    title: {
                        display: true,
                        text: 'Sales Amount'
                    },
                    ticks: {
                        callback: function(value) {
                            return formatCurrency(value);
                        }
                    }
                },
                y1: {
                    beginAtZero: true,
                    position: 'right',
                    title: {
                        display: true,
                        text: 'Quantity'
                    },
                    grid: {
                        drawOnChartArea: false
                    }
                }
            },
            plugins: {
                tooltip: {
                    callbacks: {
                        label: function(context) {
                            let label = context.dataset.label || '';
                            if (label) {
                                label += ': ';
                            }
                            if (context.datasetIndex === 0) {
                                label += formatCurrency(context.raw);
                            } else {
                                label += context.raw;
                            }
                            return label;
                        }
                    }
                }
            }
        }
    });
}

/**
 * Render top products
 */
function renderTopProducts() {
    if (!ReportsState.topProducts) return;
    
    const data = ReportsState.topProducts;
    
    // Render top products table by quantity
    const quantityTableBody = document.getElementById('topQuantityBody');
    if (quantityTableBody) {
        quantityTableBody.innerHTML = '';
        
        if (data.top_by_quantity.length === 0) {
            quantityTableBody.innerHTML = `
                <tr>
                    <td colspan="4" class="text-center">No data available</td>
                </tr>
            `;
        } else {
            data.top_by_quantity.forEach((product, index) => {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${index + 1}</td>
                    <td>${product.name}</td>
                    <td>${product.category}</td>
                    <td class="text-end">${product.quantity_sold}</td>
                `;
                quantityTableBody.appendChild(row);
            });
        }
    }
    
    // Render top products table by revenue
    const revenueTableBody = document.getElementById('topRevenueBody');
    if (revenueTableBody) {
        revenueTableBody.innerHTML = '';
        
        if (data.top_by_revenue.length === 0) {
            revenueTableBody.innerHTML = `
                <tr>
                    <td colspan="4" class="text-center">No data available</td>
                </tr>
            `;
        } else {
            data.top_by_revenue.forEach((product, index) => {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${index + 1}</td>
                    <td>${product.name}</td>
                    <td>${product.category}</td>
                    <td class="text-end">${formatCurrency(product.total_revenue)}</td>
                `;
                revenueTableBody.appendChild(row);
            });
        }
    }
}

/**
 * Render inventory status
 */
function renderInventoryStatus() {
    if (!ReportsState.inventoryStatus) return;
    
    const data = ReportsState.inventoryStatus;
    
    // Update summary cards
    document.getElementById('totalProducts').textContent = data.summary.total_products;
    document.getElementById('outOfStockCount').textContent = data.summary.out_of_stock;
    document.getElementById('lowStockCount').textContent = data.summary.low_stock;
    document.getElementById('inventoryValue').textContent = formatCurrency(data.summary.total_value);
    
    // Update threshold display
    document.getElementById('currentThreshold').textContent = data.low_stock_threshold;
    
    // Render low stock table
    const lowStockBody = document.getElementById('lowStockBody');
    if (lowStockBody) {
        lowStockBody.innerHTML = '';
        
        if (data.low_stock_products.length === 0) {
            lowStockBody.innerHTML = `
                <tr>
                    <td colspan="5" class="text-center">No low stock products</td>
                </tr>
            `;
        } else {
            data.low_stock_products.forEach(product => {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${product.barcode}</td>
                    <td>${product.name}</td>
                    <td>${product.category}</td>
                    <td class="text-warning">${product.stock_quantity}</td>
                    <td>${formatCurrency(product.price)}</td>
                `;
                lowStockBody.appendChild(row);
            });
        }
    }
    
    // Render out of stock table
    const outOfStockBody = document.getElementById('outOfStockBody');
    if (outOfStockBody) {
        outOfStockBody.innerHTML = '';
        
        if (data.out_of_stock_products.length === 0) {
            outOfStockBody.innerHTML = `
                <tr>
                    <td colspan="5" class="text-center">No out of stock products</td>
                </tr>
            `;
        } else {
            data.out_of_stock_products.forEach(product => {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${product.barcode}</td>
                    <td>${product.name}</td>
                    <td>${product.category}</td>
                    <td class="text-danger">${product.stock_quantity}</td>
                    <td>${formatCurrency(product.price)}</td>
                `;
                outOfStockBody.appendChild(row);
            });
        }
    }
    
    // Render inventory distribution chart
    renderInventoryChart(data);
}

/**
 * Render inventory distribution chart
 * @param {Object} data - Inventory status data
 */
function renderInventoryChart(data) {
    const chartCanvas = document.getElementById('inventoryChart');
    if (!chartCanvas) return;
    
    // Destroy existing chart
    if (ReportsState.charts.inventory) {
        ReportsState.charts.inventory.destroy();
    }
    
    // Prepare data
    const chartData = [
        data.summary.out_of_stock,
        data.summary.low_stock,
        data.summary.total_products - data.summary.out_of_stock - data.summary.low_stock
    ];
    
    // Create chart
    ReportsState.charts.inventory = new Chart(chartCanvas, {
        type: 'pie',
        data: {
            labels: ['Out of Stock', 'Low Stock', 'Normal Stock'],
            datasets: [{
                data: chartData,
                backgroundColor: [
                    'rgba(255, 99, 132, 0.7)',
                    'rgba(255, 205, 86, 0.7)',
                    'rgba(75, 192, 192, 0.7)'
                ],
                borderColor: [
                    'rgba(255, 99, 132, 1)',
                    'rgba(255, 205, 86, 1)',
                    'rgba(75, 192, 192, 1)'
                ],
                borderWidth: 1
            }]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            plugins: {
                legend: {
                    position: 'bottom'
                },
                tooltip: {
                    callbacks: {
                        label: function(context) {
                            const label = context.label || '';
                            const value = context.raw;
                            const total = context.dataset.data.reduce((acc, val) => acc + val, 0);
                            const percentage = Math.round((value / total) * 100);
                            return `${label}: ${value} (${percentage}%)`;
                        }
                    }
                }
            }
        }
    });
}

File: routes/reports.py
from flask import Blueprint, render_template, jsonify, request
from app import db
from models import Product, Transaction, TransactionItem
import logging
from sqlalchemy import func
from datetime import datetime, timedelta

# Create a blueprint for reports
reports_bp = Blueprint('reports', __name__, url_prefix='/reports')

@reports_bp.route('/')
def reports_page():
    """Display the reports page"""
    return render_template('reports.html')

@reports_bp.route('/sales/summary', methods=['GET'])
def sales_summary():
    """API endpoint to get sales summary data"""
    try:
        # Get filter parameters
        period = request.args.get('period', 'today')
        start_date = request.args.get('start_date')
        end_date = request.args.get('end_date')
        
        # Define date range based on period
        now = datetime.utcnow()
        if period == 'today':
            start = datetime(now.year, now.month, now.day, 0, 0, 0)
            end = datetime(now.year, now.month, now.day, 23, 59, 59)
        elif period == 'week':
            start = now - timedelta(days=7)
            end = now
        elif period == 'month':
            start = datetime(now.year, now.month, 1, 0, 0, 0)
            end = now
        elif period == 'custom':
            try:
                start = datetime.strptime(start_date, "%Y-%m-%d")
                end = datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1) - timedelta(seconds=1)
            except (ValueError, TypeError):
                return jsonify({
                    'success': False,
                    'error': 'Invalid date format. Use YYYY-MM-DD.'
                }), 400
        else:
            start = datetime(now.year, now.month, now.day, 0, 0, 0)
            end = datetime(now.year, now.month, now.day, 23, 59, 59)
        
        # Get sales data
        sales_data = db.session.query(
            func.count(Transaction.id).label('total_transactions'),
            func.sum(Transaction.total_amount).label('total_sales'),
            func.avg(Transaction.total_amount).label('average_sale')
        ).filter(
            Transaction.transaction_date.between(start, end)
        ).first()
        
        # Get payment method breakdown
        payment_methods = db.session.query(
            Transaction.payment_method,
            func.count(Transaction.id).label('count'),
            func.sum(Transaction.total_amount).label('total')
        ).filter(
            Transaction.transaction_date.between(start, end)
        ).group_by(
            Transaction.payment_method
        ).all()
        
        # Format payment methods data
        payment_data = [
            {
                'method': method, 
                'count': count, 
                'total': float(total) if total else 0
            } 
            for method, count, total in payment_methods
        ]
        
        return jsonify({
            'success': True,
            'data': {
                'period': {
                    'start': start.isoformat(),
                    'end': end.isoformat(),
                    'name': period
                },
                'sales': {
                    'total_transactions': sales_data.total_transactions if sales_data.total_transactions else 0,
                    'total_sales': float(sales_data.total_sales) if sales_data.total_sales else 0,
                    'average_sale': float(sales_data.average_sale) if sales_data.average_sale else 0
                },
                'payment_methods': payment_data
            }
        })
    except Exception as e:
        logging.error(f"Error generating sales summary: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@reports_bp.route('/sales/by-category', methods=['GET'])
def sales_by_category():
    """API endpoint to get sales data grouped by product category"""
    try:
        # Get filter parameters
        period = request.args.get('period', 'today')
        start_date = request.args.get('start_date')
        end_date = request.args.get('end_date')
        
        # Define date range based on period
        now = datetime.utcnow()
        if period == 'today':
            start = datetime(now.year, now.month, now.day, 0, 0, 0)
            end = datetime(now.year, now.month, now.day, 23, 59, 59)
        elif period == 'week':
            start = now - timedelta(days=7)
            end = now
        elif period == 'month':
            start = datetime(now.year, now.month, 1, 0, 0, 0)
            end = now
        elif period == 'custom':
            try:
                start = datetime.strptime(start_date, "%Y-%m-%d")
                end = datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1) - timedelta(seconds=1)
            except (ValueError, TypeError):
                return jsonify({
                    'success': False,
                    'error': 'Invalid date format. Use YYYY-MM-DD.'
                }), 400
        else:
            start = datetime(now.year, now.month, now.day, 0, 0, 0)
            end = datetime(now.year, now.month, now.day, 23, 59, 59)
        
        # Get sales by category
        sales_by_cat = db.session.query(
            Product.category,
            func.sum(TransactionItem.total_price).label('total_sales'),
            func.sum(TransactionItem.quantity).label('quantity_sold')
        ).join(
            TransactionItem, TransactionItem.product_id == Product.id
        ).join(
            Transaction, Transaction.id == TransactionItem.transaction_id
        ).filter(
            Transaction.transaction_date.between(start, end)
        ).group_by(
            Product.category
        ).all()
        
        # Format category data
        category_data = [
            {
                'category': category,
                'total_sales': float(total_sales) if total_sales else 0,
                'quantity_sold': quantity_sold if quantity_sold else 0
            }
            for category, total_sales, quantity_sold in sales_by_cat
        ]
        
        return jsonify({
            'success': True,
            'data': {
                'period': {
                    'start': start.isoformat(),
                    'end': end.isoformat(),
                    'name': period
                },
                'categories': category_data
            }
        })
    except Exception as e:
        logging.error(f"Error generating sales by category: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@reports_bp.route('/sales/top-products', methods=['GET'])
def top_products():
    """API endpoint to get top selling products"""
    try:
        # Get filter parameters
        period = request.args.get('period', 'today')
        start_date = request.args.get('start_date')
        end_date = request.args.get('end_date')
        limit = int(request.args.get('limit', 10))
        
        # Define date range based on period
        now = datetime.utcnow()
        if period == 'today':
            start = datetime(now.year, now.month, now.day, 0, 0, 0)
            end = datetime(now.year, now.month, now.day, 23, 59, 59)
        elif period == 'week':
            start = now - timedelta(days=7)
            end = now
        elif period == 'month':
            start = datetime(now.year, now.month, 1, 0, 0, 0)
            end = now
        elif period == 'custom':
            try:
                start = datetime.strptime(start_date, "%Y-%m-%d")
                end = datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1) - timedelta(seconds=1)
            except (ValueError, TypeError):
                return jsonify({
                    'success': False,
                    'error': 'Invalid date format. Use YYYY-MM-DD.'
                }), 400
        else:
            start = datetime(now.year, now.month, now.day, 0, 0, 0)
            end = datetime(now.year, now.month, now.day, 23, 59, 59)
        
        # Get top products by quantity sold
        top_by_quantity = db.session.query(
            Product.id,
            Product.name,
            Product.category,
            func.sum(TransactionItem.quantity).label('quantity_sold')
        ).join(
            TransactionItem, TransactionItem.product_id == Product.id
        ).join(
            Transaction, Transaction.id == TransactionItem.transaction_id
        ).filter(
            Transaction.transaction_date.between(start, end)
        ).group_by(
            Product.id, Product.name, Product.category
        ).order_by(
            func.sum(TransactionItem.quantity).desc()
        ).limit(limit).all()
        
        # Get top products by revenue
        top_by_revenue = db.session.query(
            Product.id,
            Product.name,
            Product.category,
            func.sum(TransactionItem.total_price).label('total_revenue')
        ).join(
            TransactionItem, TransactionItem.product_id == Product.id
        ).join(
            Transaction, Transaction.id == TransactionItem.transaction_id
        ).filter(
            Transaction.transaction_date.between(start, end)
        ).group_by(
            Product.id, Product.name, Product.category
        ).order_by(
            func.sum(TransactionItem.total_price).desc()
        ).limit(limit).all()
        
        # Format data
        quantity_data = [
            {
                'id': id,
                'name': name,
                'category': category,
                'quantity_sold': quantity_sold
            }
            for id, name, category, quantity_sold in top_by_quantity
        ]
        
        revenue_data = [
            {
                'id': id,
                'name': name,
                'category': category,
                'total_revenue': float(total_revenue) if total_revenue else 0
            }
            for id, name, category, total_revenue in top_by_revenue
        ]
        
        return jsonify({
            'success': True,
            'data': {
                'period': {
                    'start': start.isoformat(),
                    'end': end.isoformat(),
                    'name': period
                },
                'top_by_quantity': quantity_data,
                'top_by_revenue': revenue_data
            }
        })
    except Exception as e:
        logging.error(f"Error generating top products report: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@reports_bp.route('/inventory/status', methods=['GET'])
def inventory_status():
    """API endpoint to get current inventory status"""
    try:
        # Get low stock threshold from query params (default to 10)
        low_stock_threshold = int(request.args.get('low_stock', 10))
        
        # Get all products with stock info
        products = Product.query.all()
        
        # Calculate inventory statistics
        total_products = len(products)
        out_of_stock = sum(1 for p in products if p.stock_quantity <= 0)
        low_stock = sum(1 for p in products if 0 < p.stock_quantity <= low_stock_threshold)
        
        # Calculate total inventory value
        total_value = sum(p.price * p.stock_quantity for p in products)
        
        # Get products with low stock
        low_stock_products = Product.query.filter(
            Product.stock_quantity > 0,
            Product.stock_quantity <= low_stock_threshold
        ).all()
        
        # Get out of stock products
        out_of_stock_products = Product.query.filter(
            Product.stock_quantity <= 0
        ).all()
        
        # Format data
        low_stock_data = [product.to_dict() for product in low_stock_products]
        out_of_stock_data = [product.to_dict() for product in out_of_stock_products]
        
        return jsonify({
            'success': True,
            'data': {
                'summary': {
                    'total_products': total_products,
                    'out_of_stock': out_of_stock,
                    'low_stock': low_stock,
                    'total_value': float(total_value)
                },
                'low_stock_threshold': low_stock_threshold,
                'low_stock_products': low_stock_data,
                'out_of_stock_products': out_of_stock_data
            }
        })
    except Exception as e:
        logging.error(f"Error generating inventory status report: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500
